vpc/tool/container2/setup_container_linux.go

```diff
@@ -121,6 +140,241 @@ func DoSetupContainer(ctx context.Context, netnsfd int, assignment *vpcapi.Assig
+	}
+
+	dup, err := unix.Dup(fd)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot dup transition ns fd: %w", err)
+	}
+
+	return os.NewFile(uintptr(dup), assignment.AssignmentId), nil
+}
+
+func createTransitionNS(ctx context.Context, branchLink netlink.Link, assignment *vpcapi.AssignIPResponseV3_TransitionAssignment) (int, error) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	originalNetNS, err := netns.Get()
+	if err != nil {
+		return 0, fmt.Errorf("Could not retrieve current netns: %w", err)
+	}
+	defer originalNetNS.Close()
+
+	ns, err := netns.New()
+	if err != nil {
+		return 0, fmt.Errorf("Could not create transition ns: %w", err)
+	}
+	defer ns.Close()
+
+	handle, newHandleErr := netlink.NewHandle()
+	err = netns.Setns(originalNetNS, unix.CLONE_NEWNET)
+	if err != nil {
+		return 0, fmt.Errorf("Could not restore original network namespace: %w", err)
+	}
+
+	if newHandleErr != nil {
+		return 0, fmt.Errorf("Could not get netns handle: %w", err)
+	}
+	defer handle.Delete()
+
+	r := rand.New(rand.NewSource(time.Now().UnixNano())) // nolint: gosec
+	transitionInterfaceName := fmt.Sprintf("transition-%d", r.Intn(10000))
+	ipvlan := netlink.IPVlan{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:        transitionInterfaceName,
+			ParentIndex: branchLink.Attrs().Index,
+			Namespace:   netlink.NsFd(ns),
+			TxQLen:      -1,
+		},
+		Mode: netlink.IPVLAN_MODE_L2,
+	}
+
+	logger.G(ctx).Debugf("Adding IPVlan interface: %+v", ipvlan)
+	err = netlink.LinkAdd(&ipvlan)
+	if err != nil {
+		logger.G(ctx).WithError(err).Error("Could not add link")
+		return 0, fmt.Errorf("Cannot create link with name %q: %w", ipvlan.Name, err)
+	}
+	// If things fail here, it's fairly bad, because we've added the link to the namespace, but we don't know
+	// what it's index is, so there's no point returning it.
+	logger.G(ctx).Debugf("Added link: %+v ", ipvlan)
+	newLink, err := handle.LinkByName(ipvlan.Name)
+	if err != nil {
+		logger.G(ctx).WithError(err).Error("Could not find after adding link")
+		return 0, errors.Wrapf(err, "Cannot find link with name %s", ipvlan.Name)
+	}
+
+	err = handle.LinkSetUp(newLink)
+	if err != nil {
+		return 0, fmt.Errorf("Could not set interface %q up: %w", newLink.Attrs().Name, err)
+	}
+
+	err = addIPv4AddressAndRoutes(ctx, handle, newLink, assignment.Ipv4Address, assignment.Routes)
+	if err != nil {
+		return 0, fmt.Errorf("Could not setup transition namespace IPv4: %w", err)
+	}
+
+	/*
+		fd, err := unix.Open(transitionNamespaceDir, unix.O_CLOEXEC|unix.O_TMPFILE|unix.O_RDWR, 0755)
+		if err != nil {
+			return 0, fmt.Errorf("Cannot create tmpfile for bind: %w", err)
+		}
+	*/
+	// We do this above because the previous fd gets closed via the defer.
+	fd, err := unix.Dup(int(ns))
+	if err != nil {
+		return 0, fmt.Errorf("Could not dup netns: %w", err)
+	}
+
+	return fd, nil
+}
+
+func crossMount(ctx context.Context, src *os.File, pid1dirfd int) error {
```

Comment:
Far out.

I'm trying to come up with a potentially better way to think about this.

Because fundamentally this thing gets an fd, and creates a new namespace in a ... dir.

Maybe there is a better set of inputs and outputs for this thing to do.

Till we come up with something better though, because we control the produce and consumer, I say ship.